sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricName.java

```diff
@@ -17,63 +17,102 @@
  */
 package org.apache.beam.sdk.metrics;
 
-import static org.apache.beam.vendor.guava.v20_0.com.google.common.base.Preconditions.checkArgument;
+import static org.apache.beam.sdk.metrics.MetricUrns.USER_COUNTER_URN_PREFIX;
 
 import com.google.auto.value.AutoValue;
 import java.io.Serializable;
+import java.util.List;
+import javax.annotation.Nullable;
 import org.apache.beam.sdk.annotations.Experimental;
 import org.apache.beam.sdk.annotations.Experimental.Kind;
-import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Splitter;
 
 /**
- * The name of a metric consists of a {@link #getNamespace} and a {@link #getName}. The {@link
- * #getNamespace} allows grouping related metrics together and also prevents collisions between
- * multiple metrics with the same name.
+ * Wrapper for {@link org.apache.beam.model.pipeline.v1.MetricsApi.MonitoringInfo} URN.
+ *
+ * <p>"User" metrics (URN {@link MetricUrns#USER_COUNTER_URN_PREFIX}) are defined by a "namespace"
+ * and "name", and are the most commonly dealt with by user code, so structured constructors and
+ * accessors are provided in terms of those strings.
+ *
+ * <p>The {@link #namespace} allows grouping related metrics together and also prevents collisions
+ * between multiple metrics with the same name.
  */
 @Experimental(Kind.METRICS)
 @AutoValue
 public abstract class MetricName implements Serializable {
 
-  /** The namespace associated with this metric. */
-  public abstract String getNamespace();
+  public abstract String urn();
+
+  @Nullable private String name;
 
-  /**
-   * The namespace associated with this metric.
-   *
-   * @deprecated to be removed once Dataflow no longer requires this method.
-   */
-  @Deprecated
-  public String namespace() {
-    return getNamespace();
+  @Nullable private String namespace;
+
+  /** Parse the urn field into a name and namespace field. */
+  private boolean parseUrn(boolean raise) {
+    if (namespace != null && name != null) {
+      return true;
+    }
+    String urn = urn();
+    if (urn.startsWith(USER_COUNTER_URN_PREFIX)) {
+      urn = urn.substring(USER_COUNTER_URN_PREFIX.length());
+    } else {
+      if (raise) {
+        throw new IllegalStateException(
+            "Attempting to access namespace/name of a non-user metric: " + urn);
+      }
+      return false;
+    }
+    // If it is not a user counter, just use the first part of the URN, i.e. 'beam'
+    List<String> pieces = Splitter.on(':').splitToList(urn);
+    if (pieces.size() != 2) {
```

Comment:
@ajamato 
I believe we [allow to have colons in namespace](https://github.com/apache/beam/blob/fb425d5c4a7607d78f879e790ddebdc7bb5ea68a/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/UserMonitoringInfoToCounterUpdateTransformer.java#L106).

Reply:
Yes this might be possible. Use the last part of the split as the name, then combine everything else.

Ryan is going to move this to the labels later on, getting rid of all this parsing code anyways though

Reply:
Cool, I changed my handling to match DF, as I mentioned above, thanks.